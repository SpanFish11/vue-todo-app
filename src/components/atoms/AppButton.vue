<script setup>
const props = defineProps({
  label: { type: String, required: false, default: '' },
  customClass: { type: Array, required: false },
})
</script>

<template>
  <button :class="[customClass]">
    {{ props.label }}
  </button>
</template>
