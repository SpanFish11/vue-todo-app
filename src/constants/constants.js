export const FILTERS = Object.freeze({
  ALL: 'all',
  ACTIVE: 'active',
  COMPLETED: 'completed',
})

export const FILTER_LABELS = Object.freeze({
  all: 'All',
  active: 'Active',
  completed: 'Completed',
})
