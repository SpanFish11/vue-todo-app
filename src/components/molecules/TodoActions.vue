<script setup>
import { computed } from 'vue'
import AppButton from '@/components/atoms/AppButton.vue'
import { useStore } from 'vuex'

const store = useStore()

const filters = computed(() => store.getters.getFilters)
const remaining = computed(() => store.getters.undoneTodosCount)
const clearCompleted = () => store.dispatch('clearCompleted')
const changeVisibility = (visibility) => store.dispatch('changeVisibility', visibility)

const capitalize = (s) => s.charAt(0).toUpperCase() + s.slice(1)
</script>

<template>
  <div class="task-actions">
    <span>{{ remaining }} items left</span>
    <ul>
      <AppButton
        v-for="(val, key) in filters"
        :key="key"
        :label="capitalize(key)"
        @click="changeVisibility(key)"
        class="b"
      />
    </ul>
    <AppButton label="Clear Completed" @click="clearCompleted" />
  </div>
</template>

<style scoped lang="scss">
.task-actions {
  display: flex;
  width: 600px;
  justify-content: space-between;
  align-items: center;
  padding: 10px 0;
  font-size: 14px;
}

.b {
  margin: 5px;
}
</style>
