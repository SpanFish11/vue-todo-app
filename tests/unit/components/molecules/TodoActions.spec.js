import { beforeEach, describe, expect, it } from '@jest/globals'
import { mount, shallowMount } from '@vue/test-utils'
import TodoActions from '@/components/molecules/TodoActions.vue'
import { createStore } from 'vuex'
import { FILTER_LABELS, FILTERS } from '@/constants/constants'
import AppButton from '@/components/atoms/AppButton.vue'

describe('Given the TodoActions component', () => {
  let wrapper
  let store

  describe('when it is rendered', () => {
    beforeEach(() => {
      store = createStore({})
      store.dispatch = jest.fn()
      store.getters.undoneTodosCount = 0
      store.getters.getActiveFilter = FILTERS.ALL

      wrapper = shallowMount(TodoActions, {
        global: {
          plugins: [store],
        },
      })
    })

    it('should match the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    it.each([
      [0, 'No items left'],
      [1, '1 item left'],
      [5, '5 items left'],
    ])('should render remaining text correctly', (value, expected) => {
      store = createStore({})
      store.getters.undoneTodosCount = value

      wrapper = shallowMount(TodoActions, {
        global: {
          plugins: [store],
        },
      })

      expect(wrapper.find('span').text()).toBe(expected)
    })

    it('should show active filter', async () => {
      expect(
        wrapper
          .findAllComponents(AppButton)
          .filter((b) => b.props().label === FILTER_LABELS.all)[0]
          .props().customClass
      ).toContainEqual({ active: true })
    })
  })

  describe('when interacting with buttons', () => {
    beforeEach(() => {
      store = createStore({})
      store.dispatch = jest.fn()
      store.getters.undoneTodosCount = 0
      store.getters.getActiveFilter = FILTERS.ALL

      wrapper = mount(TodoActions, {
        global: {
          plugins: [store],
        },
      })
    })

    it('should dispatch changeActiveFilter', async () => {
      await wrapper.findByText(FILTER_LABELS.active).trigger('click')
      await wrapper.vm.$nextTick()

      expect(store.dispatch.mock.calls).toHaveLength(1)
      expect(store.dispatch).toHaveBeenCalledWith('changeActiveFilter', FILTERS.ACTIVE)
    })

    it('should dispatch clearCompleted', async () => {
      await wrapper.findByText('Clear Completed').trigger('click')
      await wrapper.vm.$nextTick()

      expect(store.dispatch.mock.calls).toHaveLength(1)
      expect(store.dispatch).toHaveBeenCalledWith('clearCompleted')
    })
  })
})
