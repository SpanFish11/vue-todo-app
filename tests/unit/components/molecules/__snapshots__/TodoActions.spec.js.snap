// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the TodoActions component when it is rendered should match the snapshot 1`] = `
<div
  class="card-body"
>
  <div
    class="d-flex justify-content-between align-items-center"
  >
    <span
      class="remaining"
    >
      No items left
    </span>
    <div>
      
      <app-button-stub
        customclass="btn-sm,me-2,[object Object]"
        label="All"
      />
      <app-button-stub
        customclass="btn-sm,me-2,[object Object]"
        label="Active"
      />
      <app-button-stub
        customclass="btn-sm,me-2,[object Object]"
        label="Completed"
      />
      
    </div>
    <app-button-stub
      customclass="btn-sm"
      label="Clear Completed"
    />
  </div>
</div>
`;
